File: code/491.non-decreasing-subsequences.java
/*
 * @lc app=leetcode id=491 lang=java
 *
 * [491] Non-decreasing Subsequences
 */

// @lc code=start

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;

class Solution {
    List<List<Integer>> result = new ArrayList<>();
    List<Integer> path = new LinkedList<>();
    public List<List<Integer>> findSubsequences(int[] nums) {
        backTracking(nums, 0);
        return result;
    }
    private void backTracking(int[] nums, int startIndex){
        if(path.size() >= 2) {
            result.add(new ArrayList<>(path));            
        }
        HashSet<Integer> hs = new HashSet<>();
        for(int i = startIndex; i < nums.length; i++){
            if(!path.isEmpty() && path.get(path.size() -1 ) > nums[i] || hs.contains(nums[i])) {
                continue;
            }
            hs.add(nums[i]);
            path.add(nums[i]);
            backTracking(nums, i + 1);
            path.removeLast();
        }
    }
}
// @lc code=end

File: code/491.non-decreasing-subsequences.go
/*
 * @lc app=leetcode id=491 lang=golang
 *
 * [491] Non-decreasing Subsequences
 */

// @lc code=start
func findSubsequences(nums []int) [][]int {
	res, path := make([][]int, 0), make([]int, 0)
	var backtracking func(start int)
	backtracking = func(start int) {
		if len(path) >= 2 {
			tmp := make([]int, len(path))
			copy(tmp, path)
			res = append(res, tmp)
		}

		hs := make(map[int]bool, len(nums))
		for i := start; i < len(nums); i++ {
			if len(path) != 0 && path[len(path)-1] > nums[i] || hs[nums[i]] {
				continue
			}

			hs[nums[i]] = true
			path = append(path, nums[i])
			backtracking(i + 1)
			path = path[:len(path)-1]
		}
	}
	backtracking(0)
	return res
}

// @lc code=end

